Implement static instruction usage histogram

ChinYikMing opened the pull request:
Copy and modify the elf.h/c from src directory for the tool to not mess up with the main  elf.h/c in src directory(code size is bigger but cleaner).

Close #34

jserv commented on tools/Makefile:
```diff
@@ -0,0 +1,71 @@
+CC ? = gcc OUT = $(CURDIR) BIN
```
Indent this file for the sake of consistency.

jserv reviewed (changes requested):
Reuse existing src/elf.[ch] as much as possible.

jserv commented on src/decode.h:
```diff
@@ -170,7 +170,7 @@
 /* IR list */
 enum {
```
Rewrite as following:
```c
enum {
#define _(inst, can_branch) rv_insn_##inst,
    RISCV_INSN_LIST
#undef _
N_RISCV_INSN_LIST
};
```

jserv commented on tools/Makefile:
```diff
@@ -0,0 +1,65 @@
+CC ?= gcc
```
Move the targets for generating tools in `mk/tools.mk` and integrate it into top-level `Makefile`.

jserv commented on tools/insn_freq_explorer.c:
```diff
@@ -0,0 +1,265 @@
+    for (int i = 0; i < hdr->e_shnum; i++) {
+        struct Elf32_Shdr *shdr = (struct Elf32_Shdr *) shdrs + i;
+
+        if (!strcmp(get_sh_type(shdr->sh_type), "PROGBITS") &&
```
Avoid comparing strings. Instead, check the values.

jserv replied:
Rewrite as following:
```c
    if ( ... COND_TO_SKIP_THIS_ITERATION ...)
        continue;
     uint32_t *exec_start_addr =
                (uint32_t *) ((uint8_t *) elf_first_byte + shdr->sh_offset);
    uint32_t *exec_end_addr =
    (uint32_t *) ((uint8_t *) exec_start_addr + shdr->sh_size);
    ...
```

jserv replied:
See https://github.com/sysprog21/rv32emu/blob/master/CONTRIBUTING.md (Avoid redundant else)

jserv commented on tools/insn_freq_explorer.c:
```diff
@@ -0,0 +1,265 @@
+{
+    int hist_bar_len = insn_freq * (max_col - used_col) / max_insn_freq;
+    for (int i = 0; i < hist_bar_len; i++) {
+        hist_bar[i] = '*';
```
Use the text rendering technique as `_mi_make_bar` (See https://github.com/microsoft/mimalloc/pull/529/files ).

jserv commented on tools/insn_freq_explorer.c:
```diff
@@ -0,0 +1,265 @@
+            uint32_t *exec_start_addr =
+                (uint32_t *) ((uint8_t *) elf_first_byte + shdr->sh_offset);
+            uint32_t *exec_end_addr =
+                (uint32_t *) ((uint8_t *) exec_start_addr + shdr->sh_size);
```
Use `uintptr_t` when possible.

jserv commented:
The result of command `tools/insn_freq_explorer build/hello.elf `
```
+---------------------------------------------+
| RV32 Target Instruction Frequency Histogram |
+---------------------------------------------+
  1. addi       0.50% [9         ] ***************************************************************************************************************************
  2. unknown    0.28% [5         ] ********************************************************************
  3. jal        0.17% [3         ] *****************************************
  4. ecall      0.11% [2         ] ***************************
  5. nop        0.06% [1         ] *************
  6. auipc      0.06% [1         ] *************
  7. beq        0.06% [1         ] *************
  8. cnop       0.06% [1         ] *************
  9. lui        0.00% [0         ] 
 10. jalr       0.00% [0         ] 
 ...
123. cjalr      0.00% [0         ] 
124. cadd       0.00% [0         ] 
125. cswsp      0.00% [0         ] 
```

We shall skip unused instructions. i.e., the one with 0%. In addition, the second `unknown` is weird.

ChinYikMing commented:
> Reuse existing src/elf.[ch] as much as possible.

I will get "incomplete definition of type" error without moving the definition of `struct Elf32_Ehdr`, `struct Elf32_Shdr` to **elf.h** header file. To preserve the encapsulation of these structures, is it suggest to implement the logic of the tool in **elf.c** file? This is the main reason that I separate new elf.[ch].

jserv commented:
> I will get "incomplete definition of type" error without moving the definition of `struct Elf32_Ehdr`, `struct Elf32_Shdr` to **elf.h** header file. To preserve the encapsulation of these structures, is it suggest to implement the logic of the tool in **elf.c** file? This is the main reason that I separate new elf.[ch].

Extend the existing src/elf.[ch] to incorporate the necessary functionalities required for implementing ELF analysis.

ChinYikMing commented:
> > I will get "incomplete definition of type" error without moving the definition of `struct Elf32_Ehdr`, `struct Elf32_Shdr` to **elf.h** header file. To preserve the encapsulation of these structures, is it suggest to implement the logic of the tool in **elf.c** file? This is the main reason that I separate new elf.[ch].
> 
> Extend the existing src/elf.[ch] to incorporate the necessary functionalities required for implementing ELF analysis.

The trivial solution is moving `struct Elf32_Ehdr`, `struct Elf32_Shdr` to **elf.h**, is it applicable?(lost encapsulation)

jserv commented:
> The trivial solution is moving `struct Elf32_Ehdr`, `struct Elf32_Shdr` to **elf.h**, is it applicable?(lost encapsulation)

`src/elf.h` in not part of public headers, so the proposed change looks fine.

qwe661234 commented:
This work looks like a static analysis, there is a more simple way to achieve this. You can just utilize a python script to parse the result of `objdump`. Because we want to keep the source code tidy and short.

jserv commented:
> This work looks like a static analysis, there is a more simple way to achieve this. You can just utilize a python script to parse the result of `objdump`. Because we want to keep the source code tidy and short.

In essence, the proposed tool is a standalone program. Given its non-essential nature to core functionalities, it is acceptable to develop a simple program that mimics the tasks of `objdump` from the GNU toolchain, eliminating the need for a runtime dependency on the GNU toolchain.

jserv commented:
@qwe661234 , By the way, how can one determine the most frequent instruction sequence during execution (in particular, CoreMark and Dhrystone), rather than focusing on a specific instruction?

qwe661234 commented:
For runtime analysis, we need to insert profiling code into emulation loop and analyze the profiling data after emulation.

jserv commented:
> For runtime analysis, we need to insert profiling code into emulation loop and analyze the profiling data after emulation.

Create an issue to track runtime analysis and profiling. There should be an explicit profiling mechanism when JIT compiler (in particular, its mixture of interpreter and/or tier compilation) .

jserv reviewed (changes requested):
Refine the git commit messages to emphasize on static instruction analysis.

jserv commented on tools/insn_freq_explorer.c:
```diff
@@ -0,0 +1,265 @@
+/*
```
Rename the file name to `tools/rv-histogram.c`. At the moment, only instruction usage is implemented. Later, we may extend for register usage.

ChinYikMing commented:
> In addition, the second `unknown` is weird.

The tool works well for other programs. I think maybe there is something wrong during the build of `hello.elf`.

jserv commented on tools/rv_histogram.c:
```diff
@@ -0,0 +1,276 @@
+#include <errno.h>
```
Add copyright notice.

jserv commented on tools/rv_histogram.c:
```diff
@@ -0,0 +1,276 @@
+static unsigned short get_win_max_col()
+{
+    struct winsize terminal_size;
+    ioctl(STDOUT_FILENO, TIOCGWINSZ, &terminal_size);
```
This tool is supposed to be portable, and you shall consider the environments lacking of full POSIX compliance. e.g., MS-Windows.

jserv commented on tools/rv_histogram.c:
```diff
@@ -0,0 +1,276 @@
+
+    return true;
+}
+
```
Remove this blank line.

jserv commented on tools/rv_histogram.c:
```diff
@@ -0,0 +1,276 @@
+    {"clwsp", 0},     {"cjr", 0},      {"cmv", 0},      {"cebreak", 0},
+    {"cjalr", 0},     {"cadd", 0},     {"cswsp", 0},
+#endif
+    {"unknown", 0}};
```
Add a comma (`,`) right after `0}`, so that clang-format can perform well.

jserv commented on tools/rv_histogram.c:
```diff
@@ -0,0 +1,276 @@
+    size_t freq;
+} rv_insn_hist_t;
+
+static rv_insn_hist_t rv_insn_str_and_freq_tbl[] = {
```
Rename this table to `rv_insn_stats`, which means the statistics of RV32 instructions including frequency and the mnemonic.

jserv commented on tools/rv_histogram.c:
```diff
@@ -0,0 +1,276 @@
+} rv_insn_hist_t;
+
+static rv_insn_hist_t rv_insn_str_and_freq_tbl[] = {
+    {"nop", 0},       {"lui", 0},      {"auipc", 0},    {"jal", 0},
```
Use [X-macro](https://en.wikipedia.org/wiki/X_macro) trick and [Stringizing](https://gcc.gnu.org/onlinedocs/cpp/Stringizing.html) to generate this table. Of course, you have to skip fuse instructions when manipulations.

jserv commented on tools/rv_histogram.c:
```diff
@@ -0,0 +1,276 @@
+#define ARRAYS_SIZE(arr) (sizeof(arr) / sizeof(arr[0]))
+
+/* 5 are fuse instructions */
+#define RISCV_INSN_ACTUAL_SIZE N_RISCV_INSN_LIST - 5
```
Rename to `RV_INSN_AVAIL_SIZE`, which emphasizes on the availability.

jserv commented:
> > the second `unknown` is weird.
> The tool works well for other programs. I think maybe there is something wrong during the build of `hello.elf`.

Create an issue to track unknown instruction appearing in statistics after this pull request is merged.

jserv commented on tools/rv_histogram.c:
```diff
@@ -0,0 +1,276 @@
+#include <sys/ioctl.h>
+#include <unistd.h>
+
+#include "../src/decode.h"
```
Don't specify `../src`. Enforce `CFLAGS` instead.

jserv commented on tools/rv_histogram.c:
```diff
@@ -0,0 +1,276 @@
+#include "../src/decode.h"
+#include "../src/elf.h"
+
+#define ARRAYS_SIZE(arr) (sizeof(arr) / sizeof(arr[0]))
```
Move the definition of `ARRAYS_SIZE` to `src/common.h`.

jserv commented on mk/tools.mk:
```diff
@@ -0,0 +1,23 @@
+OUT := build
```
You don't have to specify `OUT`.

jserv commented on tools/rv_histogram.c:
```diff
@@ -0,0 +1,376 @@
+#include <stdio.h>
```
Add copyright notice at the beginning of this file.

jserv commented on tools/rv_histogram.c:
```diff
@@ -0,0 +1,376 @@
+    size_t freq;
+} rv_insn_hist_t;
+
+#define RISCV_INSN_STATS           \
```
Do do this! You can simply use `RISCV_INSN_LIST`.

ChinYikMing replied:
But `RISCV_INSN_LIST` is concatenate the string not stringizing the string

jserv replied:
Check `src/common.h` for catenation tricks.

jserv commented on tools/rv_histogram.c:
```diff
@@ -0,0 +1,376 @@
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <sys/ioctl.h>
```
Be aware of conditional build for `_WIN32`. Both `sys/ioctl.h` and `unistd.h` might be absent.

jserv commented on Makefile:
```diff
@@ -61,7 +61,11 @@ $(OUT)/syscall_sdl.o: CFLAGS += $(shell sdl2-config --cflags)
 endif
 
-ENABLE_GDBSTUB ?= 1
+ifeq ($(filter tool,$(MAKECMDGOALS)),tool)
```
Don't do this. Minimize the necessary changes. It does not harm if gdb-stub is built with `tool` target.

jserv commented on src/elf.c:
```diff
@@ -3,6 +3,7 @@
  * "LICENSE" for information on usage and redistribution of this file.
  */
 
+#include <stdio.h>
```
This inclusion of `<stdio.h>` is not required.

jserv commented on tools/rv_histogram.c:
```diff
@@ -0,0 +1,384 @@
+    X(unknown)
+/* clang-format on */
+
+static rv_insn_hist_t rv_insn_stats[] = {
```
Remove the proposed `RISCV_INSN_STATS` macro and rewrite as following:
```c
static rv_insn_hist_t rv_insn_stats[] = {
#define _(inst, can_branch) {#inst, 0},
    RISCV_INSN_LIST
#undef _                                                                                                                                                      
};
```

Be mindful of the notation: "insn" is employed for mnemonic identifiers, while "inst" is used for parameters.

ChinYikMing replied:
Great technique for reusing `RISCV_INSN_LIST` !

jserv commented on tools/rv_histogram.c:
```diff
@@ -0,0 +1,384 @@
+
+static bool ascending_order = false;
+static const char *elf_prog = NULL;
+static const char *optstr = "a";
```
Reduce the scope of `optstr`. i.e., move it to function `parse_args`.

jserv reviewed (changes requested):
Rebase the latest `master` branch where macro-op fusion is moved out of decoder header.

jserv commented on mk/tools.mk:
```diff
@@ -0,0 +1,25 @@
+HIST_MAIN_OBJ := rv_histogram.o
+HIST_MAIN_OBJ := $(addprefix $(OUT)/, $(HIST_MAIN_OBJ))
+
+CFLAGS += -Isrc
```
Remove this line and specify `-Isrc` in the rules associated with `tools/%.c`.

jserv commented on tools/rv_histogram.c:
```diff
@@ -0,0 +1,384 @@
+#include "decode.h"
+
+/* 5 are fuse instructions */
+#define RISCV_INSN_AVAIL_SIZE N_RISCV_INSN_LIST - 5 
```
After rebasing the latest master branch, this macro no longer makes sense.

ChinYikMing committed:
Implement instruction usage histogram

This tool is designed for static analysis of instruction usage
of target program

Close #34

jserv commented:
Consider the changes below:
```diff
diff --git a/Makefile b/Makefile
index 78b0890..52a24b9 100644
--- a/Makefile
+++ b/Makefile
@@ -118,8 +118,6 @@ $(BIN): $(OBJS)
 # Tools
 include mk/tools.mk
 
-deps += $(HIST_MAIN_OBJ:%.o=%.o.d)
-
 # RISC-V Architecture Tests
 include mk/riscv-arch-test.mk
 include mk/tests.mk
@@ -166,7 +164,7 @@ endif
 endif
 
 clean:
-       $(RM) $(BIN) $(OBJS) $(HIST_BIN) $(HIST_MAIN_OBJ) $(deps) $(CACHE_OUT)
+       $(RM) $(BIN) $(OBJS) $(HIST_BIN) $(HIST_OBJS) $(deps) $(CACHE_OUT)
 distclean: clean
        -$(RM) $(DOOM_DATA) $(QUAKE_DATA)
        $(RM) -r $(OUT)/id1
diff --git a/mk/tools.mk b/mk/tools.mk
index 41f6040..11a3ac0 100644
--- a/mk/tools.mk
+++ b/mk/tools.mk
@@ -5,19 +5,19 @@ HIST_OBJS := \
        riscv.o \
        map.o \
        elf.o \
-       decode.o
-HIST_OBJS := $(addprefix $(OUT)/, $(HIST_OBJS))
+       decode.o \
+       rv_histogram.o
 
-HIST_MAIN_OBJ := rv_histogram.o
-HIST_MAIN_OBJ := $(addprefix $(OUT)/, $(HIST_MAIN_OBJ))
+HIST_OBJS := $(addprefix $(OUT)/, $(HIST_OBJS))
+deps += $(HIST_OBJS:%.o=%.o.d)
 
 $(OUT)/%.o: tools/%.c
        $(VECHO) "  CC\t$@\n"
        $(Q)$(CC) -o $@ $(CFLAGS) -Isrc -c -MMD -MF $@.d $<
 
 # GDBSTUB is diabled for excluding the mini-gdb during compilation
-$(HIST_BIN): $(HIST_OBJS) $(HIST_MAIN_OBJ)
+$(HIST_BIN): $(HIST_OBJS)
        $(VECHO) "  LD\t$@\n"
-       $(Q)$(CC) -o $@ $^ $(LDFLAGS) -D RV32_FEATURE_GDBSTUB=0
+       $(Q)$(CC) -o $@ -D RV32_FEATURE_GDBSTUB=0 $^ $(LDFLAGS)
 
 tool: $(HIST_BIN)
```

ChinYikMing committed:
Refine tool target

jserv commented:
Thank @ChinYikMing for contributing!
